### PR TITLE
Add speculative decoding telemetry and memory gating

### PR DESCRIPTION
**Pull request by aleroot**
## Proposed changes

Speculative decoding is workload and hardware-sensitive. A draft model can look good in theory but fail to pay off if acceptance is low or if the extra model pressure is too high.

The change set prepares the project for edge-aware speculative decoding by adding the two missing foundations:

1. observe speculative decoding quality/performance
2. avoid applying auxiliary-model speculation when memory pressure makes it likely to hurt

### Telemetry
It exposes whether speculative decoding is actually helping at runtime. Callers can now inspect acceptance rate, draft tokens, target verification calls, emitted tokens, and average emitted tokens per target call through GenerateCompletionInfo.speculativeDecodingTelemetry.

### Memory-Aware Gating
ChatSession now defaults speculative decoding to a memory policy based on GPU.maxRecommendedWorkingSetBytes(). Before this, enabling speculative decoding meant loading/running the main model plus draft model whenever requested. On memory-constrained systems, that can make things slower or less stable.

## Checklist

Put an `x` in the boxes that apply.

- [x] I have read the [CONTRIBUTING](https://github.com/ml-explore/mlx/blob/main/CONTRIBUTING.md) document
- [x] I have run `pre-commit run --all-files` to format my code / installed pre-commit prior to committing changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the necessary documentation (if needed)

### CODE REVIEW

File: Libraries/MLXLMCommon/Evaluate.swift [hunk 1]
```diff
@@ -756,6 +763,14 @@ public struct SpeculativeTokenIterator: TokenIteratorProtocol {
 
     // Internal metrics
     public var promptPrefillTime: TimeInterval = 0.0
+    private var telemetry = SpeculativeDecodingTelemetry()
+    public var speculativeDecodingTelemetry: SpeculativeDecodingTelemetry? {
+        telemetry.roundCount > 0 ? telemetry : nil
+    }
+
+    public mutating func discardGeneratedToken() {
+        telemetry.discardGeneratedToken()
```
Comment (davidkoski) [inline Libraries/MLXLMCommon/Evaluate.swift:774]: Should this mutate the iterator's own `tokenCount`?  Should we have a single source of truth, the telemetry?

I think the count is just informative, so not a correctness issue.

File: Libraries/MLXLMCommon/SpeculativeDecoding.swift [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+
+private func defaultSpeculativeDecodingMemoryLimit() -> Int? {
+    guard let bytes = GPU.maxRecommendedWorkingSetBytes(), bytes > 0 else {
+        return nil
+    }
+    return bytes
+}
+
+/// Runtime counters for a speculative decoding pass.
+public struct SpeculativeDecodingTelemetry: Sendable, Equatable {
```
Comment (davidkoski) [inline Libraries/MLXLMCommon/SpeculativeDecoding.swift:14]: Probably worth an issue: MTPSpeculativeTokenIterator should probably have this integrated as well.

**Comment by davidkoski** [comment]
I think this is ready to go but the GitHub web ui claims there are conflicts (SpeculativeDecodingTests.swift), however when I click on "resolve conflicts" it tells me there are zero. 😕

**Comment by davidkoski** [comment]
a rebase _does_ show a conflict, which is strange since you squashed it down to 1 commit -- the rebase and the merge that the editor use should be doing the same thing.

**Comment by davidkoski** [comment]
OK, mystery solved -- the "resolve conflicts" button does this:

```
git merge -X diff-algorithm=patience origin/main
```

which is able to cleanly resolve the merge, while `myers` (which seems to be the default strategy) cannot.  Looking at the change locally and if that is clean I will just do the resolve conflicts via the web editor.

**Comment by davidkoski** [comment]
Two tests are failing with the merge but do not fail on main:

- textOnlyPrefillChunkSizeInvariant
- `Chunked and single-pass prefill agree on logits and cache offsets`

curious because these tests are both calling `model.prepare()` with different window sizes -- a code path that isn't touched in this PR.

**Comment by aleroot** [comment]
Rebased and kept upstream’s stable-logit oracle, kept Gemma3 as smoke coverage, and preserved the telemetry assertions.

**Comment by davidkoski** [comment]
Cool, I think this unit test failure on my side is some environmental issue -- I have two checkouts and one works and the other does not.

CI underway.

**Comment by davidkoski** [comment]
Oh no, this is more TF32 business.  I had the env variable set in one of those checkouts.  OK, not related to this PR -- I will file an issue.

#357

**Review by davidkoski — APPROVED** [review]
Thank you for this improvement!
